libyul/backends/evm/ControlFlowGraphBuilder.cpp

```diff
@@ -0,0 +1,471 @@
+	// E.g. if there is a jump to a node that has the jumping node as its only entry, the nodes can be fused, etc.
+
+	return result;
+}
+
+ControlFlowGraphBuilder::ControlFlowGraphBuilder(
+	CFG& _graph,
+	AsmAnalysisInfo const& _analysisInfo,
+	Dialect const& _dialect
+):
+	m_graph(_graph),
+	m_info(_analysisInfo),
+	m_dialect(_dialect)
+{
+}
+
+StackSlot ControlFlowGraphBuilder::operator()(Literal const& _literal)
+{
+	return LiteralSlot{valueOfLiteral(_literal), _literal.debugData};
+}
+
+StackSlot ControlFlowGraphBuilder::operator()(Identifier const& _identifier)
+{
+	return VariableSlot{lookupVariable(_identifier.name), _identifier.debugData};
+}
+
+StackSlot ControlFlowGraphBuilder::operator()(Expression const& _expression)
+{
+	return std::visit(*this, _expression);
+}
+
+StackSlot ControlFlowGraphBuilder::operator()(FunctionCall const& _call)
+{
+	CFG::Operation const& operation = visitFunctionCall(_call);
+	yulAssert(operation.output.size() == 1, "");
+	return operation.output.front();
+}
+
+void ControlFlowGraphBuilder::operator()(VariableDeclaration const& _varDecl)
+{
+	yulAssert(m_currentBlock, "");
+	auto declaredVariables = _varDecl.variables | ranges::views::transform([&](TypedName const& _var) {
+		return VariableSlot{lookupVariable(_var.name), _var.debugData};
+	}) | ranges::to<vector>;
+	Stack input;
+	if (_varDecl.value)
+		input = visitAssignmentRightHandSide(*_varDecl.value, declaredVariables.size());
+	else
+		input = ranges::views::iota(0u, _varDecl.variables.size()) | ranges::views::transform([&](size_t) {
+			return LiteralSlot{0, _varDecl.debugData};
+		}) | ranges::to<Stack>;
+	m_currentBlock->operations.emplace_back(CFG::Operation{
+		std::move(input),
+		declaredVariables | ranges::to<Stack>,
+		CFG::Assignment{_varDecl.debugData, declaredVariables}
+	});
+}
+void ControlFlowGraphBuilder::operator()(Assignment const& _assignment)
+{
+	vector<VariableSlot> assignedVariables = _assignment.variableNames | ranges::views::transform([&](Identifier const& _var) {
+		return VariableSlot{lookupVariable(_var.name), _var.debugData};
+	}) | ranges::to<vector<VariableSlot>>;
+
+	yulAssert(m_currentBlock, "");
+	m_currentBlock->operations.emplace_back(CFG::Operation{
+		// input
+		visitAssignmentRightHandSide(*_assignment.value, assignedVariables.size()),
+		// output
+		assignedVariables | ranges::to<Stack>,
+		// operation
+		CFG::Assignment{_assignment.debugData, assignedVariables}
+	});
+}
+void ControlFlowGraphBuilder::operator()(ExpressionStatement const& _exprStmt)
+{
+	yulAssert(m_currentBlock, "");
+	std::visit(util::GenericVisitor{
+		[&](FunctionCall const& _call) {
+			CFG::Operation const& operation = visitFunctionCall(_call);
+			yulAssert(operation.output.empty(), "");
+		},
+		[&](auto const&) { yulAssert(false, ""); }
+	}, _exprStmt.expression);
+
+	// TODO: Ideally this would be done on the expression label and for all functions that always revert,
```

Comment:
We should have a side effect computer for that, but maybe it's still a "todo"?

Reply:
Yeah, not sure we have a computer for control flow side effects yet - and the idea was to improve it in a second step after this first version.